### PR TITLE
8294432: Add provisions to calculate hash values from MemorySegments

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes adding a _JDK-internal_ method for calculating hash codes for content in a `MemorySegment`.

The internal method uses a polynomial 32-bit hash function equivalent to `Arrays::hashCode`. The new method is almost two times faster than naïvely iterating over individual bytes for larger regions. Also, it is more lean on inlining space compared to a naïve loop.

```
Benchmark                          (ELEM_SIZE)  Mode  Cnt   Score   Error  Units
SegmentBulkHash.array                        8  avgt   30   2.645 ? 0.078  ns/op
SegmentBulkHash.array                       64  avgt   30   6.062 ? 0.171  ns/op
SegmentBulkHash.heapSegment                  8  avgt   30   4.181 ? 0.145  ns/op
SegmentBulkHash.heapSegment                 64  avgt   30  25.716 ? 1.043  ns/op
SegmentBulkHash.nativeSegment                8  avgt   30   3.939 ? 0.150  ns/op
SegmentBulkHash.nativeSegment               64  avgt   30  23.262 ? 0.694  ns/op
SegmentBulkHash.nativeSegmentJava            8  avgt   30   5.219 ? 0.183  ns/op    <- Naïve iteration
SegmentBulkHash.nativeSegmentJava           64  avgt   30  39.668 ? 1.040  ns/op    <- Naïve iteration
```

image

If internal JDK code uses this method, it will automatically benefit from future performance improvements that can be implemented once the Vector API becomes available.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294432](https://bugs.openjdk.org/browse/JDK-8294432): Add provisions to calculate hash values from MemorySegments (**Enhancement** - P3)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22364/head:pull/22364` \
`$ git checkout pull/22364`

Update a local copy of the PR: \
`$ git checkout pull/22364` \
`$ git pull https://git.openjdk.org/jdk.git pull/22364/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22364`

View PR using the GUI difftool: \
`$ git pr show -t 22364`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22364.diff">https://git.openjdk.org/jdk/pull/22364.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22364#issuecomment-2500533850)
</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java [hunk 1]
```diff
@@ -156,6 +158,94 @@ public static void copy(AbstractMemorySegmentImpl src, long srcOffset,
...
+        if (length == 0) {
+            // The state has to be checked explicitly for zero-length segments
+            segment.scope.checkValidState();
+            return 1;
+        }
+        int result = 1;
+        final long longBytes = length & ((1L << 62) - 8);
+        final long limit = fromOffset + longBytes;
+        for (; fromOffset < limit; fromOffset += 8) {
+            long val = SCOPED_MEMORY_ACCESS.getLongUnaligned(segment.sessionImpl(), segment.unsafeGetBase(), segment.unsafeGetOffset() + fromOffset, !Architecture.isLittleEndian());
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java:208]: @minborg Does it help to hoist the calls to `segment.sessionImpl()` + `segment.unsafeGetBase()` + `segment.unsafeGetOffset()` + `!Architecture.isLittleEndian()` and reusing those throughout the different strides?

Comment (minborg) [reply src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java:208]: I thought that initially as well, but keeping it like this helps the compiler select the correct unsafe code path.

File: test/jdk/java/foreign/TestSegmentBulkOperationsContentHash.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import java.lang.foreign.ValueLayout;
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+final class TestSegmentBulkOperationsContentHash {
```
Comment (mcimadamore) [inline test/jdk/java/foreign/TestSegmentBulkOperationsContentHash.java:49]: Should we add a test which checks that the result of the content hash is the same as the polynomial described in the javadoc?

File: src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java [hunk 3]
```diff
@@ -156,6 +158,94 @@ public static void copy(AbstractMemorySegmentImpl src, long srcOffset,
...
+            result = result * POWERS_OF_31[7]
+                    + ((byte) (val >>> 56)) * POWERS_OF_31[6]
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java:210]: I've manually checked the unrolling, and it seems correct

**Review by mcimadamore — APPROVED** [review]
Nice work. Let's see how often this is used and then evaluate whether to promote to public API. Let's also investigate ways to make the autovectorizer kick in for such code shapes w/o having to manually unroll :-)

File: test/jdk/java/foreign/TestSegmentBulkOperationsContentHash.java (hunk 2 above)
Comment (minborg) [reply test/jdk/java/foreign/TestSegmentBulkOperationsContentHash.java:49]: We test it is the same as `Arrays.hashCode(byte[])` which implements this scheme.

**Pull request closed without merging**